File: 150_evaluate_reverse_polish_notation/evalRPN.ts
function evalRPN(tokens: string[]): number {
  const stack: number[] = [];

  const operations = ["+", "-", "*", "/"];

  for (let token of tokens) {
    if (token.length === 1 && operations.includes(token)) {
      let b = stack.pop() as number;
      let a = stack.pop() as number;
      let operator = token;
      let value = calculate(a, b, operator);
      stack.push(value);
    } else {
      stack.push(parseInt(token));
    }
  }

  return stack.pop() || 0;
}

function calculate(a: number, b: number, operator: string): number {
  if (operator === "+") return a + b;
  if (operator === "-") return a - b;
  if (operator === "*") return a * b;
  if (operator === "/") return Math.trunc(a / b);
  return 0;
}

File: 150_evaluate_reverse_polish_notation/evalRPN.cpp
class Solution {
    public:
        int evalRPN(vector<string>& tokens) {
            stack<int> stack;
            set<string> operations = {"+", "-", "*", "/"};
    
            for (auto token : tokens) {
                if (operations.find(token) != operations.end()) {
                    int b = stack.top();
                    stack.pop();
                    int a = stack.top();
                    stack.pop();
                    int c = calculate(a, b, token);
                    stack.push(c);
                } else {
                    stack.push(stoi(token));
                }
            }
    
            return stack.top();
        }
    
    private:
        int calculate(int a, int b, string op) {
            if (op == "+") return a + b;
            if (op == "-") return a - b;
            if (op == "*") return a * b;
            if (op == "/") return a / b;
    
            return 0;
        }
    };